PhysicsTools/NanoAOD/plugins/StandAloneMuonMerger.cc

```diff
@@ -0,0 +1,137 @@
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/TrackReco/interface/TrackExtraFwd.h"
+
+#include "DataFormats/PatCandidates/interface/Muon.h"
+
+class StandAloneMuonMerger : public edm::stream::EDProducer<>
+{
+public:
+  explicit StandAloneMuonMerger(const edm::ParameterSet &);
+  ~StandAloneMuonMerger() {}
+
+//   static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
+
+private:
+  
+  virtual void produce(edm::Event &, const edm::EventSetup &) override;
+  
+  edm::EDGetTokenT<reco::TrackCollection> inputTrackStandAlone_;
+  edm::EDGetTokenT<reco::TrackCollection> inputTrackStandAloneUpdatedAtVtx_;
+  edm::EDGetTokenT<std::vector<pat::Muon>> inputMuons_;
+  edm::EDPutTokenT<reco::TrackCollection> outputTrack_;
+  edm::EDPutTokenT<edm::ValueMap<bool>> outputUpdatedAtVtxMap_;
+  edm::EDPutTokenT<edm::ValueMap<bool>> outputMuonUpdatedAtVtxMap_;
+
+};
+
+
+StandAloneMuonMerger::StandAloneMuonMerger(const edm::ParameterSet &iConfig)
+
+{
+  inputTrackStandAlone_ = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("standAlone"));
+  inputTrackStandAloneUpdatedAtVtx_ = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("standAloneUpdatedAtVtx"));
+  inputMuons_ = consumes<std::vector<pat::Muon>>(iConfig.getParameter<edm::InputTag>("muons"));
+  
+  outputTrack_ = produces<reco::TrackCollection>();
+  outputUpdatedAtVtxMap_ = produces<edm::ValueMap<bool>>("updatedAtVtx");
+  outputMuonUpdatedAtVtxMap_ = produces<edm::ValueMap<bool>>("muonUpdatedAtVtx");
+}
+
+// ------------ method called for each event  ------------
+void StandAloneMuonMerger::produce(edm::Event &iEvent, const edm::EventSetup &iSetup)
+{
+  using namespace edm;
+  
+  Handle<reco::TrackCollection> tracksStandAlone;
+  iEvent.getByToken(inputTrackStandAlone_, tracksStandAlone);
+
+  Handle<reco::TrackCollection> tracksStandAloneUpdatedAtVtx;
+  iEvent.getByToken(inputTrackStandAloneUpdatedAtVtx_, tracksStandAloneUpdatedAtVtx);
+  
+  Handle<std::vector<pat::Muon>> muons;
+  iEvent.getByToken(inputMuons_, muons);
+  
+  reco::TrackCollection tracksOut;
+  
+  // fill collection of disambiguated standalone muon tracks plus value map indicating whether
+  // the selected track is updated at vertex or not
+  
+  std::vector<bool> updatedAtVtxV;
+  updatedAtVtxV.reserve(tracksStandAlone->size());
+  
+  for (auto const &standAlone : *tracksStandAlone) {
+    auto const extraIdx = standAlone.extra().key();
+    
+    const reco::Track *track = &standAlone;
+    bool updatedAtVtx = false;
+    
+    // replicate logic in GlobalMuonProducer, take the updatedAtVtx track if it exists and has
+    // the same eta sign as the original, otherwise take the original
+    for (auto const &standAloneUpdatedAtVtx : *tracksStandAloneUpdatedAtVtx) {
+      if (standAloneUpdatedAtVtx.extra().key() == extraIdx) {
+        const int etaFlip1 = (standAloneUpdatedAtVtx.eta() * standAlone.eta()) < 0 ? -1 : 1;
```

Comment:
why is this an int?

Reply:
The code is deliberately directly copy-pasted from the GlobalMuonProducer https://github.com/cms-sw/cmssw/blob/c86fcb3bfafb1089a873558d087b89d432bc0c23/RecoMuon/GlobalMuonProducer/src/GlobalMuonProducer.cc#L124-L132